projectq/cengines/_optimize.py

```diff
@@ -114,88 +111,365 @@ def _get_gate_indices(self, idx, i, IDs):
+            self._l[qubitids[j]] = new_list
+
+    def _get_erase_boolean(self, idx, qubitids, commandidcs, inverse_command, apply_commutation):
+        """
+        Determines whether inverse commands should be cancelled
+        with one another. i.e. the commands between the pair are all
+        commutable for each qubit involved in the command.
+        """
+        erase = True
+        # We dont want to examine qubit idx because the optimizer 
+        # has already checked that the gates between the current 
+        # and mergeable gates are commutable (or a commutable list).
+        commandidcs.pop(qubitids.index(idx)) # Remove corresponding 
+        # position of command for qubit idx from commandidcs
+        qubitids.remove(idx) # Remove qubitid representing the current 
+        # qubit in optimizer
+        x=1
+        for j in range(len(qubitids)):
+            # Check that any gates between current gate and inverse
+            # gate are all commutable
+            this_command = self._l[qubitids[j]][commandidcs[j]]
+            future_command = self._l[qubitids[j]][commandidcs[j]+x]
+            while (future_command!=inverse_command):
+                if apply_commutation==False:
+                    # If apply_commutation turned off, you should 
+                    # only get erase=True if commands are next to
+                    # eachother on all qubits. i.e. if future_command
+                    # and inverse_command are not equal (i.e. there
+                    # are gates separating them), you don't want 
+                    # optimizer to look at whether the separating gates 
+                    # are commutable.
+                    return False
+                if (this_command.is_commutable(future_command)==1):
+                    x+=1
+                    future_command = self._l[qubitids[j]][commandidcs[j]+x]
+                    erase = True
+                else:
+                    erase = False
+                    break
+            if (this_command.is_commutable(future_command)==2):
+                new_x = self._check_for_commutable_circuit(this_command, future_command, qubitids[j], commandidcs[j], 0)
+                if(new_x>x):
+                    x=new_x
+                    future_command = self._l[qubitids[j]][commandidcs[j]+x]
+                    erase=True
+                else:
+                    erase=False
+                    break
+        return erase
+
+    def _get_merge_boolean(self, idx, qubitids, commandidcs, merged_command, apply_commutation):
+        """
+        To determine whether mergeable commands should be merged
+        with one another. i.e. the commands between them are all
+        commutable, for each qubit involved in the command. It does
+        not check for the situation where commands are separated by
+        a commutable list. However other parts of the optimizer 
+        should find this situation.
```

Comment:
Could you give an example for such a situation? I am not sure I understand that limitation.

Reply:
I think I just meant that this function does not take into account commutable_circuits. I can't think of a situation where this would be relevant actually so I'll just get rid of this comment.